### ISSUE TITLE
Automatically switch sites to the GA4 dashboard view on October 1

### ISSUE THREAD

**Issue by nfmohit** [google/site-kit-wp#6549]
## Feature Description

~~Note: This is a time-sensitive issue. This shall only be executed to be released as a part of the [Site Kit update/release going right after UA has a zero-data state](https://docs.google.com/document/d/14pCNmdPBqAJ-RfPhbHXfaMhGaPK-n-7F-6Sgujd0XCQ/edit#heading=h.ocr7bivzb5b2).~~ The changes outlined here will be date-based, i.e. developed as we go, but will be visible to users on or after October 1 2023.

Around 90 days after the UA cut-off, Site Kit will no longer show any UA data anymore naturally. At this point, for the sites that have GA4 connected, we want to switch them automatically to the GA4 dashboard view. To accomplish this, we will need to make a few small changes like the following:

- The `isGA4DashboardView` selector (being implemented in #6541) should be updated so that if GA4 is connected, it should unconditionally return `true`. Since Site Kit will automatically switch to the GA4 dashboard view at this point, this selector returning true will not require us to change the logic immediately and give us a generous amount of time to [remove the selector and the UA module overall](https://docs.google.com/document/d/17RKIrdm5PBK8j6rP0GV_2QlztSLXqYQX8ikz3kQmrAo/edit?resourcekey=0-njOU7bn1dhvi5Yp2jeX5zw#heading=h.lmmhmll0a7xd).
- ~~Since we will automatically switch sites to GA4 dashboard view, we will no longer need to prompt users to switch over. Hence, the “Switch to GA4 Dashboard View” `BannerNotification` (being added in #6544) should be removed, along with the `shouldPromptGA4DashboardView` selector in the `analytics-4` datastore.~~ This will be done as a part of #6786.
- ~~Since at this point there will be no way to switch back to the UA dashboard view, the dashboard view toggle in Analytics settings (being added in #6547) should be removed.~~ This will be done as a part of #6786.
- Copy of the in-widget warnings about UA cut-off (added in #6557) should be updated according to the [Figma design](https://www.figma.com/file/0UCIE3hPeiu4NYdpAODASd/GA4-Updates?node-id=559%3A6540&t=ckvN59ELFbj9QQWN-0).

[Here](https://docs.google.com/document/d/14pCNmdPBqAJ-RfPhbHXfaMhGaPK-n-7F-6Sgujd0XCQ/edit#heading=h.ocr7bivzb5b2) is the relevant part in the design doc.

---------------

_Do not alter or remove anything below. The following sections will be managed by moderators only._

## Acceptance criteria

* If the current date is October 1 2023 or later, and the `ga4Reporting` feature flag is enabled:
  * The `isGA4DashboardView` selector in the `analytics` datastore should be modified to always return `true` if GA4 is connected.
  * The copy/text of the in-widget warnings about UA cut-off (added in #6557) should be updated according to the content in the ["Post-July" section of the Figma design](https://www.figma.com/file/0UCIE3hPeiu4NYdpAODASd/GA4-Updates?node-id=559%3A6540&t=ckvN59ELFbj9QQWN-0).
  * The "Switched to GA4" notification banner (developed in #6558) should be made visible to the users (i.e. added to `<BannerNotifications />`).
    * In terms of priority, the banner should be displayed just below the `ModuleRecoveryAlert` banner. It should appear above `ZeroDataStateNotifications`.
  * The `Dashboard view` setting in Analytics module settings (Site Kit -> Settings) (both view and edit) should not be displayed.
* The current date can be obtained using the `getReferenceDate()` selector in the `core/user` store. It is being enhanced as a part of #6782 so that an optional custom reference date can be set on page load for testing purposes.

## Implementation Brief

- In `assets/js/modules/analytics/datastore/settings.js`:
  - Update the `isGA4DashboardView()` selector so that it also returns `true` if the following conditions are met:
    - The `ga4Reporting` feature flag is enabled.
    - Current date (obtainable via `getReferenceDate()` selector in the `core/user` store) is October 1 2023 or later.
    - GA4 is connected (obtainable via `isModuleConnected( 'analytics-4' )` selector in the `core/modules` store).
- In `assets/js/modules/analytics/components/common/UACutoffWarning.js`:
  - If the current date (obtainable via `getReferenceDate()` selector in the `core/user` store) is October 1 2023 or later:
    - Update the text/copy of the `<SettingsNotice />` component's `notice` prop according to the Analytics widget area notices in [this screen](https://www.figma.com/file/0UCIE3hPeiu4NYdpAODASd/GA4-Updates?node-id=559-6540&t=jNNTthhHHYISOLjL-0).
    - The text referred to in the design as of writing the IB is: `No fresh data to display. Universal Analytics stopped collecting data on July 1. To resume collecting Analytics data, set up Google Analytics 4.`.
    - The `Learn more` link and `Set up Google Analytics 4` CTA should persist as of today.
- In `/assets/js/components/notifications/SwitchedToGA4Banner.js` (being added in #6558):
  - Update the conditions to show the `BannerNotification` such that besides the existing conditions, it is also checked if the current date (obtainable via `getReferenceDate()` selector in the `core/user` store) is October 1 2023 or later. 
- In `assets/js/components/notifications/BannerNotifications.js`:
  - Import the `/assets/js/components/notifications/SwitchedToGA4Banner.js` component (being added in #6558).
  - Place it after the `<ModuleRecoveryAlert />` component, only if the `ga4Reporting` feature flag is enabled.
- In `assets/js/modules/analytics/components/settings/SettingsForm.js`:
  - Update the conditions under which the `Dashboard view` setting fields group is displayed so that besides the current conditions, it is also checked if the current date (obtainable via `getReferenceDate()` selector in the `core/user` store) is October 1 2023 or later - the group should _not_ be displayed if so.
- In `assets/js/modules/analytics/components/settings/SettingsView.js`:
  - Update the conditions under which the `Dashboard view` setting meta view is displayed so that besides the current conditions, it is also checked if the current date (obtainable via `getReferenceDate()` selector in the `core/user` store) is October 1 2023 or later - the meta view should _not_ be displayed if so.

### Storybook

- Add Storybook story variant for `UACutOffWarning` to display its updated copy after UA-removal (October 1).
- Add Storybook story variants for Analytics `SettingsEdit` & `SettingsView` showing its difference of state after UA cut-off (July 1) and UA-removal (October 1).

### Test Coverage

- In `assets/js/modules/analytics/datastore/settings.test.js`:
  - Update test cases for the `isGA4DashboardView()` selector to reflect the above changes.
- In `assets/js/modules/analytics/components/common/UACutoffWarning.test.js`:
  - Update test cases for the component to reflect the above changes.
- Fix any other failing tests.
- Add VRT scenarios for the updated stories.

## QA Brief

- Setup Site Kit with the `ga4Reporting` feature flag enabled and connect Analytics (both GA4 & UA).
- Set the Dashboard View setting to Universal Analytics.
  - Note that you may want to use the following snippet in order to set the dashboard view _without_ causing the GA4 feature tour to be dismissed, which is what happens when updating the view via the Settings page.
  - `googlesitekit.api.set('modules', 'analytics', 'settings', { dashboardView: 'universal-analytics' })`
- Using the Tester plugin, set the reference date to `2023-10-01` or later. Alternatively use the following snippet to set the date via the JS console while using the plugin.
  - `googlesitekit.data.dispatch('core/user').setReferenceDate('2023-10-01')`
- Verify that the Dashboard View setting is overridden to be GA4. This can be confirmed by running the following selector which should now return `true`, as well as by viewing the Dashboard and checking that the GA4 versions of the widgets are displayed.
  - `googlesitekit.data.select('modules/analytics').isGA4DashboardView()`
- Verify that the Dashboard View section is no longer present on the Analytics Settings edit and view pages.
- Verify that the Dashboard View indicator chip is no longer present on the Dashboard.
- Verify the text in the UA Cutoff Warning notices is updated _as per the IB_ i.e. the text should be:
  - _No fresh data to display. Universal Analytics stopped collecting data on July 1. To resume collecting Analytics data, set up Google Analytics 4._
    - This is consistent with the pre-Oct 1st text which is _"Your data is stale because Universal Analytics stopped collecting data on July 1, 2023."_ The text in Figma contains the phrasing _"_No fresh data to display. Universal Analytics property stopped collecting data on July 1."_ which is grammatically slightly off and inconsistent.
  - To display the notice, GA4 needs to be in a disconnected state while UA is connected. The easiest way to achieve this is to delete the `googlesitekit_analytics-4_settings` option from the DB.
- Verify that the "Switched to GA4" notification banner appears when the Dashboard View setting is overridden via setting the date.
  - Refer to the first four QAB steps in https://github.com/google/site-kit-wp/issues/6558.
  - However, rather than manually switching over to the GA4 Dashboard View in the second step, instead run the JS snippet mentioned above to set the Dashboard View to Universal Analytics (use the snippet to avoid dismissing the feature tour, as the notification won't display when the tour has been dismissed).
  - Then, when logged in as the second admin, ensure the date is set to `2023-10-01` or later, in order to override the Dashboard View to GA4.

## Changelog entry

* Automatically switch users to GA4 dashboard on October 1, 2023 (when no more Universal Analytics data will be available).

**Comment by techanvil** [comment]
Have moved this back to AC to determine whether to develop this behind a feature flag instead of trying to target a specific release.

**Comment by nfmohit** [comment]
Thank you, @techanvil!

I have updated the ACs to mention the addition of the Banner Notification developed in #6558 to `<BannerNotifications />` as a part of this issue.

Other than that, @aaemnnosttv @tofumatt do you folks have any thoughts regarding @techanvil's idea of using some sort of a flag (e.g. a new feature flag) to develop/test this ahead of time, and fine-control the activation date instead of relying on people updating their Site Kit version? ([Slack discussion](https://10up.slack.com/archives/CBKKQEBR9/p1678811817307509))

Thanks!

**Comment by techanvil** [comment]
Thanks @nfmohit. To be honest, I'd have thought the existing feature flag system would be the most pragmatic choice here, but I'm not sure what if any alternative we currently have in terms of infrastructure that lets us update some data remotely...

**Comment by aaemnnosttv** [comment]
@techanvil I agree that some kind of controllable flag would be best here rather than making this release-dependent. Certainly a feature flag could be used for this condition, but I don't think there are that many moving pieces here that would justify it, particularly on the server side, nor do we need the the gradual rollout. We could implement the flag "internally" and essentially calculate it rather than sourcing it as a persistent remote flag as we usually do, which isn't something we've done before but is something we could potentially consider.

How about using a dismissed item instead (e.g. `uaDashboard`)? These are persistent across resets and should be sufficiently configurable for testing. We would then automatically set the item so long as UA and GA4 were connected. This would only be needed as long as the UA infra was still around and we already have the infra in place to get and set it.

**Comment by nfmohit** [comment]
> One more thing I wanted to highlight from the current AC is that it seems to be inconsistent with the description which says that we only want to do the automatic transition if GA4 is connected. This nuance isn't represented in the current first point:
> 
> > * The `isGA4DashboardView` selector in the `analytics-4` datastore should be modified to always return `true`.
> 
> Also, this selector is currently part of the `analytics` store, not GA4.

Thank you for spotting this, @aaemnnosttv! I have updated this.

**Comment by techanvil** [comment]
> @techanvil I agree that some kind of controllable flag would be best here rather than making this release-dependent. Certainly a feature flag could be used for this condition, but I don't think there are that many moving pieces here that would justify it, particularly on the server side, nor do we need the the gradual rollout. We could implement the flag "internally" and essentially calculate it rather than sourcing it as a persistent remote flag as we usually do, which isn't something we've done before but is something we could potentially consider.
> 
> How about using a dismissed item instead (e.g. `uaDashboard`)? These are persistent across resets and should be sufficiently configurable for testing. We would then automatically set the item so long as UA and GA4 were connected. This would only be needed as long as the UA infra was still around and we already have the infra in place to get and set it.

Hmm. Well, if we're going to be determining the value ourselves, does it need to be a persisted value at all?

Also, seeing as the intention here is to toggle some behaviour [after September 30, 2023](https://docs.google.com/document/d/14pCNmdPBqAJ-RfPhbHXfaMhGaPK-n-7F-6Sgujd0XCQ/edit#heading=h.ocr7bivzb5b2), if we're not going to do this remotely or via a release, why don't we simply treat this as another date-dependent set of functionality?

**Comment by aaemnnosttv** [comment]
> Hmm. Well, if we're going to be determining the value ourselves, does it need to be a persisted value at all?

@techanvil not necessarily, but it could be useful to have it leverage existing infra.

> if we're not going to do this remotely or via a release, why don't we simply treat this as another date-dependent piece of functionality?

Given that the changes only seem to be relevant for the front end it could be date-based using the current reference date. The only problem with that is that we don't currently have an easy way to control this for QA ahead of time.

On its face, there's many ways this could be done, but I think aiming for something which is straightforward to QA would be a good guiding principal.

Conceptually, I feel like this fits well with dismissed items since the UA dashboard is kind of being "dismissed" permanently. The mechanism for setting the dismissal could be triggered by conditions using the reference date, and for QA we could use a console command to trigger that dismissal. If we wanted to go for a bit cleaner of an approach, we could use a feature flag but have the enabled state of that featured flag controlled by a dismissed item. That way it could still be toggled via the tester plugin but would be controlled by the plugin itself.

That's my 2c. Thoughts @felixarntz ?

**Comment by techanvil** [comment]
Hi @aaemnnosttv, thanks for the detailed explanation. I've certainly no objection to the direction you've suggested here, and am interested to hear @felixarntz' thoughts too.

One alternative possibility I would like to suggest though, a suggestion that could keep the front-end code a bit simpler, while also allowing ease of QA, would be to implement this as a simple check against the reference date, and then to facilitate QAing, we could add a feature to the helper plugin that allows the reference date to be overridden so it's set at page load time. That would allow simple QAing of this as well as other reference-date related features...

**Comment by aaemnnosttv** [comment]
@techanvil that sounds reasonable to me and something I had considered at some point as well. The one issue that might be a problem there is the availability of that value which is currently only exposed through the data store.

We'd need to introduce a mechanism for providing this on page load as well as make that addition in the tester plugin but that would be fine. I think this much is worth doing anyways as we'll have some date-based logic in here regardless. Would you please open a new issue for this part? Then the AC would only need the added detail here about when the change should take effect, and this issue would depend on the new one.

**Comment by tofumatt** [comment]
Just weighing in on this discussion: I too think a "reference date" feature in the tester plugin that can be used for this feature makes the most sense (over a feature flag or dismissed item).

Coding this with a date in mind is straightforward and communicates the intent in the code, and then for QA/development/testing allowing either the date used as the "cutoff" or the date used as "today" in the calculation seems good. (I suspect it'll be the former, but either works for me.) 👍🏻

(I'm going to unassign myself from this as I don't have anything else to add at the moment, but please assign me again if you want another look/opinion/action from me 🙂 )

**Comment by aaemnnosttv** [comment]
Thanks @tofumatt, I think we can split this into two issues going forward - that is essentially repurposing this one to be specific to changes that should take effect on/after the date, and another issue to actually remove things that will no longer be relevant/needed in a subsequent release.

As for the date-based logic, I think we'll still want to abstract this into a common selector/etc to avoid repetitive and probably more verbose use of using `getReferenceDate` inline. I'll open a new issue for the changes related to the reference date infra and then we can make this issue dependent on that.

Assigning this one back to @nfmohit to iterate on per this direction.

**Comment by techanvil** [comment]
Thanks @aaemnnosttv, sorry it took me a little while to get round to this one, a few other things got in the way (notably the GTE launch blockers). I would have raised an issue as asked, but looks like you're taking care of this now, so thanks for that!

**Comment by nfmohit** [comment]
Thank you for the collaboration here, folks!

> I think we can split this into two issues going forward - that is essentially repurposing this one to be specific to changes that should take effect on/after the date, and another issue to actually remove things that will no longer be relevant/needed in a subsequent release.

I have updated the ACs for this issue to scope changes that should take effect on/after September 1 2023. I've created #6786 to be "release-based" (on or right after September 1 2023) to remove the UA bits.

CC: @aaemnnosttv

**Comment by nfmohit** [comment]
> In addition to the condition of the current date, I think we also want to require the ga4Reporting flag? It's probably already there, but in terms of requirements, I think everything here would still be contingent on the feature flag

Updated, thanks!

> Would you please clarify what you're referring to by "Setup GA4 widget area error banner"? Are you referring to [these](https://www.figma.com/file/0UCIE3hPeiu4NYdpAODASd/GA4-Updates?node-id=559-6797&t=bn7HV0VZZ5zdCIX9-4)?

Yes. I've updated this section to be clearer. Also, the ACs previously mentioned updating the GA4 Activation `BannerNotification`. But that is not needed because that's already handled in #6538. Thus, I've removed this part from the ACs.

Thank you, @aaemnnosttv!

**Comment by aaemnnosttv** [comment]
Thanks @nfmohit – AC ✅

**Comment by aaemnnosttv** [comment]
@nfmohit yes, there is no point in displaying that toggle when there is only one usable version of the dashboard. The only reason we might keep it (present but disabled) would be to support a persistent inline notice in its place that the dashboard only supports GA4 going forward?

**Comment by aaemnnosttv** [comment]
@nfmohit I updated the dates here, is there anything left to add or update on your end?

**Comment by nfmohit** [comment]
Thank you, @aaemnnosttv!

I was waiting to see if @marrrmarrr has any recommendations on the approach to take regarding [this](https://github.com/google/site-kit-wp/issues/6549#issuecomment-1490045272) (either completely remove or disable it with notice) in Slack [here](https://10up.slack.com/archives/CFFRMC5DE/p1680197787927229).

As soon as there is feedback, I'll update the ACs and the IB for that part. Let me know if you have any thoughts about this. Thanks!

**Comment by nfmohit** [comment]
@aaemnnosttv Just a heads-up that I've updated the ACs to include not rendering the `Dashboard view` setting on/after October 1, 2023.

**Comment by techanvil** [comment]
Hey @nfmohit, the IB is looking great. I've made a couple of very small tweaks which you can see in the edit history. Please review these to make sure it's aligned with your thinking.

Additionally - I'm wondering if we should add some Storybook story variants to capture these changes. I'm thinking we could add one for `UACutoffWarning` to show the new copy, and potentially variants for Analytics `SettingsEdit` / `SettingsView` to show them after the cutoff date, with the Dashboard View no longer displayed - although these wouldn't be so interesting to look at, they could help with test coverage if VRT scenarios are included. What do you think?

**Comment by nfmohit** [comment]
> Hey @nfmohit, the IB is looking great. I've made a couple of very small tweaks which you can see in the edit history. Please review these to make sure it's aligned with your thinking.
> 
> Additionally - I'm wondering if we should add some Storybook story variants to capture these changes. I'm thinking we could add one for `UACutoffWarning` to show the new copy, and potentially variants for Analytics `SettingsEdit` / `SettingsView` to show them after the cutoff date, with the Dashboard View no longer displayed - although these wouldn't be so interesting to look at, they could help with test coverage if VRT scenarios are included. What do you think?

Thank you for the kind review and feedback, @techanvil. I've updated the IB to include Storybook updates. Let me know if it looks good.

**Comment by techanvil** [comment]
Thanks @nfmohit! I made another minor tweak, as there is already a `UACutoffWarning` story for the UA cut-off version, and we just need to add the UA-removal version.

IB :white_check_mark:

**Comment by techanvil** [comment]
Note: Have moved this back to the EB, as the additional point added to the AC is no longer needed and has been removed (see the description edit history for the date of this comment). This is no longer needed due to the updated AC for https://github.com/google/site-kit-wp/issues/6914.

**Comment by mohitwp** [comment]
### QA Update ⚠️ 

@techanvil When date is set to 1 Oct, 2023. Then under Analytics settings `The Dashboard view` settings is not displaying but UA settings are still visible. Will we continue to display UA settings when the dashboard view permanently switches to GA4?

![image](https://github.com/google/site-kit-wp/assets/94359491/c31cbe97-0eee-4618-b673-887a5df29192)

**Comment by techanvil** [comment]
Hi @mohitwp, thanks for asking! We will indeed be removing this section, however this particular issue is simply about changing the dashboard view setting, which doesn't affect the rendering of UA settings.

The section you've highlighted will be removed via a separate issue, https://github.com/google/site-kit-wp/issues/6786.

**Comment by mohitwp** [comment]
QA Update : Created new issue for Dashboard sharing - View only dashboard related issue > #7417
